// @svelte-web-fonts/google/dist/google-module.js
export default class GoogleModule {
    static isItalic(variant) {
        return variant.includes("italic");
    }
    static getFamilyDefinition(definition) {
        const { family } = definition;
        const weights = GoogleModule.getFamilyWeights(definition);
        return `${family.replace(/ /g, "+")}:${weights}`;
    }
    static getFamilyWeights(definition) {
        const { variants } = definition;
        const someItalic = variants.some(GoogleModule.isItalic);
        const pattern = someItalic ? "ital,wght" : "wght";
        const weights = variants
            .map((variant) => {
            if (someItalic) {
                const italvalue = GoogleModule.isItalic(variant) ? 1 : 0;
                const weight = variant.replace("italic", "");
                return `${italvalue},${weight}`;
            }
            return variant;
        })
            .join(someItalic ? ";" : ",");
        return `${pattern}@${weights}`;
    }
    static applyParamToUrl(url, param, value) {
        return `${url}&${param}=${encodeURIComponent(value)}`;
    }
    static getStylesheet(fonts, { display = "auto", subsets = [], effect = "none", text = "", } = {}) {
        const _fonts = Array.isArray(fonts) ? fonts : [fonts];
        const families = _fonts
            .map(GoogleModule.getFamilyDefinition)
            .map((family, index) => {
            const prefix = index === 0 ? "?" : "&";
            return `${prefix}family=${family}`;
        })
            .join("");
        let url = `${GoogleModule.baseUrl}${families}`;
        if (display !== "auto") {
            url = GoogleModule.applyParamToUrl(url, "display", display);
        }
        if (subsets.length > 0) {
            const subsetsString = subsets.join(",");
            url = this.applyParamToUrl(url, "subset", subsetsString);
        }
        if (text) {
            url = this.applyParamToUrl(url, "text", text);
        }
        if (effect !== "none") {
            url = this.applyParamToUrl(url, "effect", effect);
        }
        return url;
    }
}
GoogleModule.baseUrl = `https://fonts.googleapis.com/css2`;
